### ISSUE TITLE
FLT_EPSILON is too small for RotatedRect

### ISSUE THREAD

**Issue by takotakot** [opencv/opencv#12455]
##### System information (version)
- OpenCV => 3.4.1 (master seems to be same)
- Operating System / Platform => cygwin 64 bit
- Compiler => g++ (gcc 7.3.0)

##### Detailed description

In some (very rare) condition, "Error: Assertion failed" happens at:
https://github.com/opencv/opencv/blob/da7e1cfb495be131fe7d4668b163cd1cd75e7670/modules/core/src/types.cpp#L154
`CV_Assert(abs(vecs[0].dot(vecs[1])) / (cv::norm(vecs[0]) * cv::norm(vecs[1])) <= FLT_EPSILON);`
because `FLT_EPSILON` is too small to compare.

##### Steps to reproduce

I made a reproducible example:
https://github.com/takotakot/opencv_debug/tree/0a5d37dc2cc4aef22a33012bdcbb54597ae852a1
.
If we have `Point2d` rectangle, using `RotatedRect_2d` can eliminate the problem.

Part of the code:

```.cpp
    cv::Mat points = (cv::Mat_<double>(10, 2) <<
                    1357., 1337.,
                    1362., 1407.,
                    1367., 1474.,
                    1372., 1543.,
                    1375., 1625.,
                    1375., 1696.,
                    1377., 1734.,
                    1378., 1742.,
                    1382., 1801.,
                    1372., 1990.);

    cv::PCA pca_points(points, cv::Mat(), CV_PCA_DATA_AS_ROW, 2);

    cv::Point2d p1(564.45, 339.8819), p2, p3;
    p2 = p1 - 1999 * cv::Point2d(pca_points.eigenvectors.row(0));
    p3 = p2 - 1498.5295 * cv::Point2d(pca_points.eigenvectors.row(1));

    cv::RotatedRect(p1, p2, p3);
```

##### Plans

I have some plans:

1. Multiple 2, 4 or some value to FLT_EPSILON
1. Make another constructor using `Point2d` for `Point2f` (and `Vec2d` for `Vec2f` etc. inside)
   Note 1: If we use `DBL_EPSILON`, same problem may occur.
   Note 2: If we only have `Point2f` rectangle, we cannot avoid assertion.
1. Calcurate the angle between two vectors and introduce another assersion.

I want to create PR for solving this issue. But I want some direction.

**Comment by berak** [comment]
please see [the previous discussion about it](https://github.com/opencv/opencv/issues/12221) (duplicate)

**Comment by takotakot** [comment]
@berak Thank you for commenting.
I've already read the issue.
#12221 is about intersections and #12222 is merged.
The point that "double is better" is the same, but this is not not a duplicate bug.

**Comment by alalek** [comment]
Please try something like this:
```
CV_Assert(std::fabs(vecs[0].ddot(vecs[1])) <= 0.017 /*cos(89deg)*/ * cv::norm(vecs[0]) * cv::norm(vecs[1]));
```

Changes:
- `.ddot`
- `FLT_EPSILON` ==> `cos(89 deg)` (almost 90 degrees)
- avoid division

**Comment by takotakot** [comment]
For my environment:

- `.ddot`
- `fabs`
- avoid division

does not solve this bug.

- `FLT_EPSILON` to `1.7`

is good.

My `Point2d` version:

```
ddot: -3.63797880709171e-10
norm0: 1999
norm1: 1498.5295
prod: 2995560.4705
1.21445680797239e-16
```
1.21445680797239e-16 is smaller than `FLT_EPSILON`.

My `Point2f` version:
```
ddot: -0.406243801116943
norm0: 1999.00007932771
norm1: 1498.52945663236
prod: 2995560.50268301
1.35615288275128e-07
OpenCV(3.4.1) Error: Assertion failed (abs(vecs[0].dot(vecs[1])) / (norm(vecs[0]) * norm(vecs[1])) <= 1.19209289550781250000e-7F)
```

1.35615288275128e-07 is a bit larger than `FLT_EPSILON`.

A patch that change `FLT_EPSILON` to `std::sqrt(FLT_EPSILON)` or some value is accepted?
`double` version is better?

**Comment by takotakot** [comment]
> This magic value doesn't have any sense.

Yes, I konw that. But 89 degree does not have sense neither. How about 4*EPS?
I will calculate a^2 + b^2 - c^2 using norm.

**Comment by takotakot** [comment]
Using `a^2 = vecs[0].ddot(vecs[0])` and `vec2 = p2 - p0`:

```
norm0: 1999.00014033596
norm1: 1498.52945663236
norm2: 2498.31772711755
a^2 + b^2 - c^2: 0.627828244119883
(a^2 + b^2 - c^2) / c^2: 1.00587846477474e-07
fabs((a^2 + b^2) /c^2 - 1): 1.0058784649658e-07
```

0.627828244119883 is relatively small. So, `fabs( a^2 + b^2 - c^2 ) <= FLT_EPSILON * c^2` passes.

With `double` version, `fabs((a^2 + b^2) /c^2 - 1)` is just 0.

#### Proposal:

`CV_Assert((std::fabs(vecs[0].ddot(vecs[1])) <= FLT_EPSILON  * (cv::norm(vecs[0]) * cv::norm(vecs[1]))) || (std::fabs(vecs[0].ddot(vecs[0]) + vecs[1].ddot(vecs[1]) - vec2.ddot(vec2)) <= FLT_EPSILON * vec2.ddot(vec2)));`

This method is better when:

- corner case (I reported)
- width/height is relatively large/small than the other

Replace `FLT_EPSILON` with `FLT_EPSILON * 2` may be better, too.

**Comment by takotakot** [comment]
#### Estimate calculation (quantization) error:

Define
![image](https://user-images.githubusercontent.com/10368878/45276350-abb00d00-b4fc-11e8-888e-12385b4b1ccc.png)
and
ac+bd=0
Squared order of `epsilon`s are denoted as `o( epsilon ^2)`
Define
![image](https://user-images.githubusercontent.com/10368878/45276452-57f1f380-b4fd-11e8-9081-3dd9340076f2.png)


Inner product:

![image](https://user-images.githubusercontent.com/10368878/45276494-9d162580-b4fd-11e8-8c0b-b3d2df4d975f.png)
Thus 
![image](https://user-images.githubusercontent.com/10368878/45276502-b0c18c00-b4fd-11e8-912b-d2c071d291fd.png)

Norm:
![image](https://user-images.githubusercontent.com/10368878/45276524-d484d200-b4fd-11e8-8881-95d8efd28c72.png)

![image](https://user-images.githubusercontent.com/10368878/45276538-f3836400-b4fd-11e8-98d4-9b61cba04dfe.png)

In this way, `(4 + alpha) | epsilon |` (where alpha is some small value) seems to be sufficient.


#### Estimate cancellation of significant digits:

If three points
![image](https://user-images.githubusercontent.com/10368878/45277074-e74cd600-b500-11e8-8300-3ea412cab1b4.png)
 (k=0, 1, 2) are given. We can calculate v_0 = p_1 - p_0, v_1 = p_2 - p_1 .
Assume | x_k | and | y_k | are relatively larger than | a | etc.

For computer calcuration, quantization error exists ex.:
![image](https://user-images.githubusercontent.com/10368878/45277066-cedcbb80-b500-11e8-8d96-22d82d4b8664.png)


And we have cancellation of significant digits as follows:
![image](https://user-images.githubusercontent.com/10368878/45276720-eb77f400-b4fe-11e8-84f2-7e6588901b1a.png)

Then

![image](https://user-images.githubusercontent.com/10368878/45277127-3e52ab00-b501-11e8-9f16-9dc05c9118ad.png)


if | a | is relatively smaller than | b | we can ignore this.

For convenience define:

![image](https://user-images.githubusercontent.com/10368878/45276767-3eea4200-b4ff-11e8-97aa-86e611f95850.png)

We can estimate:

![image](https://user-images.githubusercontent.com/10368878/45276793-5295a880-b4ff-11e8-8c38-41a31f2215b3.png)

#### proposal 2

![image](https://user-images.githubusercontent.com/10368878/45276800-63deb500-b4ff-11e8-9dcc-401990aa4063.png)

**Comment by alalek** [comment]
Feel free to prepare PR onto 3.4 branch.

Assumption:
```
x' = x * (1 + eps), where eps < FLT_EPS
```
is not always true due calculations in caller function (they may use cos/sin approximations with 100+ ULP).

**Comment by takotakot** [comment]
Thank you.

Yes. There are many factors.
I've created #12523.

**Comment by takotakot** [comment]
Closed via #12523.
